README.md
plug-socket
===========

Simple plug.dj WebSocket EventEmitter library.

[![NPM](https://nodei.co/npm/plug-socket.png?downloads)](https://nodei.co/npm/plug-socket) [![Greenkeeper badge](https://badges.greenkeeper.io/miniplug/plug-socket.svg)](https://greenkeeper.io/)

## Usage

```javascript
const plugSocket = require('plug-socket')
const authToken = '(...)' // get one by GET-ing https://plug.dj/_/auth/token

let socket = plugSocket(authToken)
// events will be fired on "socket" for every incoming plug.dj message
socket.on('chat', msg => log(`<${msg.un}>  ${msg.message}`))
socket.on('userJoin', msg => log(`  * ${msg.un} joined the room`))
socket.on('plugMaintenanceAlert', () => log('#ded soon… ×.×'))
```

## API

### let socket = plugSocket(?authToken)

Sets up a WebSocket connection to plug.dj. If the auth token is given, it also
sends an "auth" message once the connection is open. Otherwise, you'll have to
send that yourself.

`socket` is a [WebSocket](https://github.com/websockets/ws) connection instance
with a few extra methods and a bunch of extra events.

### socket.auth(authToken)

Sends an auth token. You should only call this once, and only if you did not
pass one to the `plugSocket()` call.

You can obtain an auth token by logging in to plug.dj using something like
[plug-login](https://github.com/miniplug/plug-login), or by manually
sending a GET request to https://plug.dj/_/auth/token.

```javascript
const plugSocket = require('plug-socket')

// Using `plug-login`'s authToken option:
const plugLogin = require('plug-login')
plugLogin(myEmail, myPassword, { authToken: true }).then((result) => {
  const sock = plugSocket(result.token)
})

// Or manually, with a cookie stored in `mySessionCookie`:
const got = require('got')
got('https://plug.dj/_/auth/token', {
  json: true,
  // Make
  headers: { cookie: mySessionCookie }
}).then((response) => {
  const authToken = response.body.data[0]
  const sock = plugSocket(authToken)
})
```

### socket.chat(message)

Sends a chat message to the current room. Make sure to join a room first by
sending a POST request to https://plug.dj/_/rooms/join:

```javascript
got.post('https://plug.dj/_/rooms/join', {
  json: true,
  headers: {
    cookie: mySessionCookie,
    'content-type': 'application/json'
  },
  body: JSON.stringify({
    slug: 'my-room-slug'
  })
}).then((response) => { /* joined! */ })
```

### Events

Aside from the standard WebSocket events, `plug-socket` also emits different
events for all plug.dj message types. These are:

```javascript
[ "ack", "advance", "ban", "banIP", "chat", "chatDelete", "djListCycle"
, "djListLocked", "djListUpdate", "earn", "sub", "cash", "gift", "floodChat"
, "floodAPI", "friendRequest", "friendAccept", "gifted", "grab", "killSession"
, "modBan", "modAddDJ", "modRemoveDJ", "modMoveDJ", "modMute", "modSkip"
, "modStaff", "nameChanged", "nameChangedRoom", "notify", "playlistCycle"
, "plugMaintenance", "plugMaintenanceAlert", "plugMessage", "plugUpdate"
, "rateLimit", "roomNameUpdate", "roomDescriptionUpdate", "roomWelcomeUpdate"
, "roomMinChatLevelUpdate", "skip", "userJoin", "userLeave", "userUpdate"
, "vote" ]
```

Plug.dj events receive two arguments, `param` and `slug`. `param` is *usually*
an object, or `undefined` for some events. The `slug` parameter contains the
current room slug or "dashboard". When you switch rooms, sometimes you'll keep
receiving a few events from your previous room, so the `slug` parameter allows
you to filter those. You won't have to care for it if your app doesn't switch
rooms much.

```javascript
socket.on('chat', (param, slug) => {
  log(`${slug}: receiving`, param)
})
```

You can also handle _every_ plug.dj event by adding an "action" listener:

```javascript
socket.on('action', (type, param, slug) => {
  // `type` is one of the events listed above.
  log(`${slug}: receiving a "${type}" event with`, param)
})
```

Most events are documented in more detail in the [PlugCommunity Documentation][plugcommunity docs]
repository.

## Tests

Tests use `mocha`. All tests depend on plug.dj being online and reachable, so
you might get test failures if it's slow, or in maintenance mode, or shut down
for a few months.

## License

[MIT][license]

[plugcommunity docs]: https://github.com/plugcommunity/documentation/tree/master/api/events/backend_events/
[license]: ./LICENSE


CHANGELOG.md
# plug-socket change log

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](http://semver.org/).

## 1.0.0 / 03 May 2018
* Drop browser support (useless anyway).
* Use a Rollup build like other miniplug packages.


test.js
var fetch = require('node-fetch')
var assert = require('assert')
var http = require('http')
var WebSocket = require('ws')
var login = require('plug-login')
var socket = require('./')

var room = 'plug-socket-test'

var token
var user
describe('plug.dj', function () {
  this.timeout(30000)

  it('is reachable', function () {
    return fetch('https://plug.dj/').then(function (response) {
      return response.text()
    }).then(function (body) {
      if (body.indexOf('<title>maintenance') !== -1)
        throw new Error('plug.dj is currently in maintenance mode.')
    })
  })

  it('can connect as guest and gets a valid auth token', function () {
    return login.guest({ authToken: true }).then(function (result) {
      assert.ok(result)
      assert.ok(result.token)
      token = result.token
    })
  })
})

describe('plug-socket', function () {
  this.timeout(30000)
  var s

  beforeEach(function () {
    assert.ok(token, 'has an auth token')
  })

  afterEach(function () {
    if (s) s.close()
  })

  it('can connect and authenticate with an auth token', function (done) {
    s = socket(token)
    s.once('ack', function (param) {
      assert.strictEqual(param, '1')
      done()
    })
  })

  it('can connect without an initial auth token', function (done) {
    s = socket()
    s.on('ack', assert.fail)
    setTimeout(function () {
      s.removeListener('ack', assert.fail)

      s.on('ack', function (param) {
        assert.strictEqual(param, '1')
        done()
      }).auth(token)
    }, 1000)
  })

  it('emits events for chat', function (done) {
    var user = { id: 342546, username: 'test' }
    var testMsg = 'This is plug-socket speaking!'
    s = socket()
    s.on('chat', function (msg) {
      if (msg.uid === user.id) {
        assert.strictEqual(msg.message, testMsg)
        done()
      }
    })
    s.once('open', function () {
      s.onmessage({
        data: JSON.stringify([ {
          a: 'chat',
          p: {
            cid: user.id + '-' + Date.now(),
            message: testMsg,
            sub: 0,
            uid: user.id,
            un: user.username
          },
          s: room
        } ])
      })
    })
  })

  it('emits "action" events for each individual message', function (done) {
    s = socket()

    var calledEarn = false
    var calledGift = false
    var called = {}

    s.on('earn', function () {
      calledEarn = true
    })
    s.on('gift', function () {
      calledGift = true
    })
    s.on('action', function (action, param, slug) {
      called[action] = true
      assert.ok(param)
      assert.ok(slug)
    })

    s.once('open', function () {
      s.onmessage({
        data: JSON.stringify([
          { a: 'earn', p: { xp: 1000 }, s: 'dashboard' },
          { a: 'gift', p: { uid: -1 }, s: 'tastycat' }
        ])
      })

      assert.ok(calledEarn)
      assert.ok(calledGift)
      assert.ok(called.earn)
      assert.ok(called.gift)

      done()
    })
  })

  it('disconnects if a message has not been received for some time', function (done) {
    this.timeout(2000)

    var server = http.createServer().listen()
    var wsserver = new WebSocket.Server({ server: server })
    s = socket(null, {
      timeout: 300,
      url: 'ws://localhost:' + server.address().port
    })

    s.onclose = function (event) {
      assert.equal(event.code, 3001)

      server.close()
      done()
    }
  })
})


test/test.js
var request = require('request')
var assert = require('assert')
var WebSocket = require('ws')
var login = require('plug-login')
var socket = require('../')

var room = 'plug-socket-test'

var jar = request.jar()
var token
var user
describe('plug.dj', function () {
  this.timeout(30000)

  it('is reachable', function (done) {
    request('https://plug.dj/', function (e, _, body) {
      if (e)
        throw e
      if (body.indexOf('<title>maintenance') !== -1)
        throw new Error('plug.dj is currently in maintenance mode.')
      done()
    })
  })

  it('can connect as guest', function (done) {
    login.guest({ jar: jar }, function (e, result) {
      if (e) return done(e)
      assert.ok(result)
      done()
    })
  })

  it('gives a valid auth token', function (done) {
    login.getAuthToken({ jar: jar }, function (e, authToken) {
      if (e) return done(e)
      assert.ok(authToken)
      token = authToken
      done()
    })
  })
})

describe('plug-socket', function () {
  this.timeout(30000)

  it('can connect and authenticate with an auth token', function (done) {
    socket(token).once('ack', function (param) {
      assert.strictEqual(param, '1')
      done()
    })
  })

  it('can connect without an initial auth token', function (done) {
    var s = socket()
    s.on('ack', assert.fail)
    setTimeout(function () {
      s.removeListener('ack', assert.fail)

      s.on('ack', function (param) {
        assert.strictEqual(param, '1')
        done()
      }).auth(token)
    }, 1000)
  })

  it('emits events for chat', function (done) {
    var user = { id: 342546, username: 'test' }
    var testMsg = 'This is plug-socket speaking!'
    var s = socket()
    s.on('chat', function (msg) {
      if (msg.uid === user.id) {
        assert.strictEqual(msg.message, testMsg)
        done()
      }
    })
    s.emit('message', JSON.stringify([
      {
        a: 'chat',
        p: {
          cid: user.id + '-' + Date.now(),
          message: testMsg,
          sub: 0,
          uid: user.id,
          un: user.username
        },
        s: room
      }
    ]))
  })
})


src/index.js
import WebSocket from 'ws'

const WS_URL = 'wss://ws-prod.plug.dj:443/socket'
const WS_ORIGIN = 'https://plug.dj'

const WSSTATE_OPEN = 1
const HEARTBEAT_TIMEOUT = 25 * 1000

export default function socket (authToken, options) {
  const wsUrl = options && options.url || WS_URL
  const wsOrigin = options && options.origin || WS_ORIGIN
  const heartbeatTimeout = options && options.timeout || HEARTBEAT_TIMEOUT

  const ws = new WebSocket(wsUrl, { origin: wsOrigin })

  const queue = []
  let heartbeat

  function gotHeartbeat () {
    if (heartbeat) clearTimeout(heartbeat)
    heartbeat = setTimeout(ontimedout, heartbeatTimeout)
  }

  function onmessage (event) {
    gotHeartbeat()

    if (event.data === 'h') {
      return null
    }

    const actions = JSON.parse(event.data)
    if (!Array.isArray(actions)) {
      return null
    }

    actions.forEach((data) => {
      // Action shape:
      // { a: action, p: param, s: slug }
      ws.emit(data.a, data.p, data.s)
      ws.emit('action', data.a, data.p, data.s)
    })
  }

  ws.sendMessage = function sendMessage (action, param) {
    if (ws.readyState === WSSTATE_OPEN) {
      ws.send(JSON.stringify({
        a: action,
        p: param,
        t: Math.floor(Date.now() / 1000)
      }))
    } else {
      queue.push({ action: action, param: param })
    }
    return ws
  }

  /**
   * Send all queued messages.
   */
  function onopen () {
    gotHeartbeat()
    queue.forEach((message) => {
      ws.sendMessage(message.action, message.param)
    })
  }

  function onclose () {
    if (heartbeat) clearTimeout(heartbeat)
  }

  /**
   * When we haven't received a heartbeat for some time, the connection might
   * have stopped working.
   */
  function ontimedout () {
    ws.close(3001, 'Timed out: did not receive heartbeat from plug.dj')
  }

  ws.auth = function auth (param) {
    return ws.sendMessage('auth', param)
  }
  ws.chat = function chat (param) {
    return ws.sendMessage('chat', param)
  }

  ws.onmessage = onmessage
  ws.onopen = onopen
  ws.onclose = onclose

  if (authToken) {
    ws.auth(authToken)
  }

  return ws
}
